File: backend/src/polyhear/models/project.py
"""Pydantic models for project data."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class StatusLevel(str, Enum):
    """Status level enum."""

    PASS = "pass"
    FAIL = "fail"
    UNKNOWN = "unknown"
    SKIPPED = "skipped"


class WorkingStatus(BaseModel):
    """Git working directory status."""

    clean: bool = True
    staged: int = 0
    modified: int = 0
    untracked: int = 0


class RemoteStatus(BaseModel):
    """Remote tracking status."""

    ahead: int = 0
    behind: int = 0
    tracking: str | None = None


class BuildStatus(BaseModel):
    """Build status from log file."""

    status: StatusLevel = StatusLevel.UNKNOWN
    exit_code: int | None = None
    duration_seconds: float | None = None
    timestamp: datetime | None = None
    message: str | None = None


class TestStatus(BaseModel):
    """Test status from log file."""

    status: StatusLevel = StatusLevel.UNKNOWN
    passed: int = 0
    failed: int = 0
    skipped: int = 0
    duration_ms: int | None = None
    timestamp: datetime | None = None
    failures: list[dict[str, str]] = Field(default_factory=list)


class LastCommit(BaseModel):
    """Last commit information."""

    hash: str
    message: str
    time: str  # Relative time like "5 minutes ago"
    author: str | None = None


class ProjectSummary(BaseModel):
    """Progress summary for a project."""

    completed: str | None = None
    working: str | None = None
    failed: str | None = None
    todos: int = 0
    fixmes: int = 0


class ProjectError(BaseModel):
    """Error information for a project."""

    code: str
    message: str
    suggestion: str
    details: str | None = None


class WorktreeInfo(BaseModel):
    """Git worktree information."""

    path: str
    branch: str
    commit_hash: str
    is_bare: bool = False
    is_detached: bool = False


class Project(BaseModel):
    """Project overview for list view."""

    id: str
    name: str
    path: str
    branch: str
    status: WorkingStatus
    remote: RemoteStatus
    build: BuildStatus
    test: TestStatus
    last_commit: LastCommit | None = None
    summary: ProjectSummary
    error: ProjectError | None = None
    parent: str | None = None  # Parent project name for worktrees
    worktrees: list["Project"] = Field(default_factory=list)


class ProjectDetail(Project):
    """Detailed project information."""

    recent_commits: list[LastCommit] = Field(default_factory=list)
    uncommitted_changes: list[dict[str, str]] = Field(default_factory=list)
    worktree_info: WorktreeInfo | None = None


class ProjectCreate(BaseModel):
    """Request model for creating a project."""

    name: str
    path: str
    group: str | None = None
    parent: str | None = None
    build_log: str | None = None
    test_log: str | None = None


class ProjectListResponse(BaseModel):
    """Response model for project list."""

    projects: list[Project]
    updated_at: datetime

File: frontend/src/types/project.ts
export type StatusLevel = 'pass' | 'fail' | 'unknown' | 'skipped'

export interface WorkingStatus {
  clean: boolean
  staged: number
  modified: number
  untracked: number
}

export interface RemoteStatus {
  ahead: number
  behind: number
  tracking?: string
}

export interface BuildStatus {
  status: StatusLevel
  exit_code?: number
  duration_seconds?: number
  timestamp?: string
  message?: string
}

export interface TestStatus {
  status: StatusLevel
  passed: number
  failed: number
  skipped: number
  duration_ms?: number
  timestamp?: string
  failures: { name: string; message: string }[]
}

export interface LastCommit {
  hash: string
  message: string
  time: string
  author?: string
}

export interface ProjectSummary {
  completed?: string
  working?: string
  failed?: string
  todos: number
  fixmes: number
}

export interface ProjectError {
  code: string
  message: string
  suggestion: string
  details?: string
}

export interface Project {
  id: string
  name: string
  path: string
  branch: string
  status: WorkingStatus
  remote: RemoteStatus
  build: BuildStatus
  test: TestStatus
  last_commit?: LastCommit
  summary: ProjectSummary
  error?: ProjectError
  parent?: string
  worktrees: Project[]
}

export interface ProjectDetail extends Project {
  recent_commits: LastCommit[]
  uncommitted_changes: { status: string; path: string }[]
}

export interface ProjectCreate {
  name: string
  path: string
  group?: string
  parent?: string
  build_log?: string
  test_log?: string
}

export interface ProjectListResponse {
  projects: Project[]
  updated_at: string
}

export interface UISettings {
  theme: string
  refresh_interval: number
  max_commits: number
}

export interface SettingsResponse {
  ui: UISettings
  summary: {
    scan_todo: boolean
    todo_patterns: string[]
  }
  llm: {
    enabled: boolean
    provider: string
  }
}
